installinator/src/mock_peers.rs

```diff
@@ -0,0 +1,504 @@
+}
+
+#[derive(Clone, Debug)]
+#[allow(dead_code)]
+enum MockResponse {
+    // Once the actions run out, this times out.
+    Response(Vec<ResponseAction>),
+    NotFound { after: Duration },
+    Forbidden { after: Duration },
+}
+
+#[derive(Clone, Debug)]
+struct ResponseAction {
+    after: Duration,
+    count: usize,
+}
+
+#[cfg(test)]
+mod proptest_helpers {
+    use proptest::prelude::*;
+    use test_strategy::Arbitrary;
+
+    use super::*;
+
+    impl MockPeersUniverse {
+        pub(crate) fn strategy(
+            max_peer_count: usize,
+        ) -> impl Strategy<Value = Self> {
+            let artifact_strategy = prop_oneof![
+                // Don't try shrinking the bytes inside the artifact -- their individual values don't
+                // matter.
+                99 => prop::collection::vec(any::<u8>().no_shrink(), 0..4096),
+                // Make it not very unlikely that the artifact is empty.
+                1 => Just(Vec::new()),
+            ];
+
+            // We can assume without loss of generality that content is fetched from peers in
+            // ascending IPv6 order. In other words, the addresses themselves aren't relevant beyond
+            // being unique identifiers. This means that this code can use a BTreeMap rather than a
+            // fancier structure like an IndexMap.
+            let peers_strategy = prop::collection::btree_map(
+                any::<Ipv6Addr>(),
+                any::<MockResponse_>(),
+                0..max_peer_count,
+            );
+
+            // Any u32 is a valid bitmap. If there are fewer peers than bits in the bitmap, the
+            // higher-order bits will be ignored.
+            let attempt_bitmaps_strategy =
+                prop::collection::vec(any::<AttemptBitmap>(), 1..16);
+
+            (artifact_strategy, peers_strategy, attempt_bitmaps_strategy)
+                .prop_map(
+                    |(artifact, peers, attempt_bitmaps): (
+                        Vec<u8>,
+                        BTreeMap<Ipv6Addr, MockResponse_>,
+                        Vec<AttemptBitmap>,
+                    )| {
+                        let artifact = Bytes::from(artifact);
+                        let peers = peers
+                            .into_iter()
+                            .map(|(peer, response)| {
+                                let response = response.into_actual(&artifact);
+                                (
+                                    peer,
+                                    MockPeer {
+                                        artifact: artifact.clone(),
+                                        response,
+                                    },
+                                )
+                            })
+                            .collect();
+                        Self::new(artifact, peers, attempt_bitmaps)
+                    },
+                )
+        }
+    }
+
+    #[derive(Debug, Arbitrary)]
+    enum MockResponse_ {
```

Comment:
Is this like a proptest thing? Why have structs and enums that look nearly identical to existing ones? Couldn't we just have these derives on `MockResponse`, rather than creating a distinct `MockResponse_`?

(This is less of a suggestion, more of a question; I'm not super familiar with proptest)

Reply:
So the `MockResponse_` data structures are basically identical to `MockResponse`, except they contain a [`proptest::sample::Index`](https://altsysrq.github.io/rustdoc/proptest/latest/proptest/sample/struct.Index.html) to indicate indexes rather than a `usize`.

The alternative would be to use a generic here -- I'm not entirely sure if that works with the derive macros I'm using here, but I can look!

As to why this is required, I'll write up a thing about some of the theory behind proptest at some point. The tl;dr is that we want to try and do our best to not use `prop_flat_map` (monadic bind). There are several ways to work around it and using indexes that get resolved later is one of them.